File: LeetCode/123BestTimeToBuyAndSellStockIII/BestTimeToBuyAndSellStockIII.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        // Bidirectional DP -> Time:O(n), Space:O(n)
        vector<int> left(prices.size(), 0), right(prices.size(), 0);
        int n = prices.size(), minPrice = prices[0], maxPrice = prices[n-1], maxProfit = 0;
        for (int i = 1; i < n; i++) {
            minPrice = min(minPrice, prices[i]);
            left[i] = max(left[i-1], prices[i] - minPrice);
        }
        for (int i = n-2; i > -1; i--) {
            maxPrice = max(maxPrice, prices[i]);
            right[i] = max(right[i+1], maxPrice - prices[i]);
        }
        for (int i = 0; i < n; i++) {
            maxProfit = max(maxProfit, left[i]+right[i]);
        }
        return maxProfit;
    }
};

File: LeetCode/123BestTimeToBuyAndSellStockIII/BestTimeToBuyAndSellStockIII.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # Bidirectional DP -> Time:O(n), Space:O(n)
        left = [0]*len(prices)
        right = [0]*len(prices)
        min_price = prices[0]
        for i in range(1, len(prices)):
            min_price = min(prices[i], min_price)
            left[i] = max(left[i-1], prices[i] - min_price)
        max_price = prices[-1]
        for i in range(len(prices)-2, -1, -1):
            max_price = max(prices[i], max_price)
            right[i] = max(right[i+1], max_price - prices[i])
        max_profit = 0
        for i in range(len(prices)):
            max_profit = max(max_profit, left[i]+right[i])
        return max_profit
